plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/XenServerGuru.java

```diff
@@ -98,7 +98,12 @@ public VirtualMachineTO implement(VirtualMachineProfile vm) {
         if (userVmVO != null) {
             HostVO host = hostDao.findById(userVmVO.getHostId());
             if (host != null) {
-                to.setVcpuMaxLimit(MaxNumberOfVCPUSPerVM.valueIn(host.getClusterId()));
+                // Max cpu cannot be more than host capability
+                if (host.getCpus() < MaxNumberOfVCPUSPerVM.valueIn(host.getClusterId()))
+                    to.setVcpuMaxLimit(host.getCpus());
```

Comment:
I think host.getCpus() get the actual CPU on the physical host, but xenserver host is capable of allocating virtual CPUs more than the actual count.
Here is the list of configuration limits for XS 7.1 version. https://docs.citrix.com/en-us/xenserver/7-1/system-requirements/configuration-limits.html
It says "Virtual CPUs per VM (Linux)" 32. So we might need to set the value according to this.

Reply:
I tested this with a host with 8 CPUs and it failed trying to allocate 16 on XCP-ng 7.6

Reply:
I'll do some testing on Xenserver

Reply:
I tested XCP-NG 7.6 and Xenserver 7.6, I was unable to start a VM with more vCPUs than pCPUs.
![XCP-NG](https://user-images.githubusercontent.com/49917670/98325437-7d082100-1ff7-11eb-92c3-025412e7da43.png)
![Xenserver](https://user-images.githubusercontent.com/49917670/98325446-81343e80-1ff7-11eb-819c-2aa7567b2065.png)

Reply:
Seems like overcommit on xenserver was removed according to this article @harikrishna-patnala https://support.citrix.com/article/CTX236977

Reply:
It might also depend on the supported limits on guest OS. but setting the max vcpu limit less than or equal to host cpus does not give complete leverage to scale the VM. Also this setting was there since long time so I think this is something to do with OS type of XCP ng version limitation.

Reply:
@Spaceman1984 got it, thanks for the article and yes it seems over committing CPU is disabled.
Do you think we have to also provide some global config "vcpu-unrestricted" to allow overcommit of CPU like xenserver does this way "xe vm-param-set uuid=<uuid> platform:vcpu-unrestricted=true". This gives admins to choose whether to overcommit or not.

Otherwise If we have to go with by setting only actual pCPUs, we need to remove this configuration parameter "xen.vm.vcpu.max" as we dont need it now.

Reply:
I kept the global var in to allow things like if you have a host with 32 pCPUs but want to set a lower max vCPU limit for a VM for the purpose of dynamic scaling.
As for the ability to force overcommit - This is not reccommended in the article and I would think not best practice, that's why I would vote no.

Reply:
Okay LGTM.

Reply:
please put a log message which explans why we are setting that value, otherwise user may think why a different value is set other than "xen.vm.vcpu.max"

Reply:
Alright, awesome, on it.